Add beginner's guide to testing agents blog post (#16107)
PostHog/posthog.com: contents/blog/testing-ai-agents.md

---
date: 2026-03-31
title: "A beginner's guide to testing AI agents"
author:
  - radu-raicea
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: Engineering
tags:
  - Engineering
  - LLM analytics
---

Agents are becoming increasingly present in our lives and work. We use many agents, directly or indirectly, and as engineers, we're increasingly building them too.

Not too long ago, creating software meant writing the business logic and then writing tests, at least to the extent that the software engineer was concerned. Now, the work of engineers building agents has shifted.

This shift has made what we're building much harder to test properly. Before, when business logic was almost always deterministic, you could write a solid set of unit, integration, and UI tests to confidently defend against bugs and regressions. Now, many engineers building agents seem to have lost that testing ability and are resorting to manually testing their agent or depending on users to report bugs, bugs that could have been caught earlier. It seems almost comical to have to argue for automated tests, but it feels necessary.

> The good news is that, as with traditional software testing, there is a way to do it, both when building your agent and when improving it. Let's discover it together by going through the life of an agent.

Reviewer: Refocusing this intro on the old way of testing vs agent testing as I think that is the strongest bit and it's kind of hidden. 
```suggestion
Software testing used to be relatively straightforward.

You wrote business logic, defined expected outputs, and verified them with unit, integration, and UI tests. With the same input, your system behaved the same way every time.

AI agents changed this.

Instead of fixed logic, we now build systems powered by models that are probabilistic. The same input can produce different outputs. Behavior depends on prompts, context, tool calls, and external data. The "correct" answer is fuzzy.

And yet, many teams still try to test AI agents the old way, or worse, not at all. Manual testing and bug reports quietly become the default.

The good news is that testing hasn’t become impossible, it’s just changed. Like before, there is a way to systematically test and improve agent behavior. It just requires a different mental model.

In this post, we’ll walk through what testing AI agents well looks like by following the life of an agent, from first build to production and beyond.
```

## The birth of the agent

> It all begins with the classic "Hello World", though now it's wrapped inside an SDK for creating agents, and suddenly we're not saluting the world, but rather being a helpful assistant.

Reviewer: ```suggestion
It all begins with the classic "Hello World", though now it's wrapped inside an SDK for creating agents, and we're not greeting the world, but rather being a helpful assistant.
```

3. Look at traces that have been flagged as problematic via a support ticket
4. If you've shipped a new feature, look for traces that are using that feature, for example by looking at the [traces that have called a specific tool](/docs/llm-analytics/tools) belonging to that feature
5. If your LLM observability platform supports it, look at traces that have a [negative sentiment](/docs/llm-analytics/sentiment) associated with them
6. Also if supported, look at outlier traces or small clusters from the [clusters generated for your traces](/docs/llm-analytics/clusters)

The PostHog teams working on AI features and agents have a weekly ritual called Traces Hour, where they look at and review traces that have been marked for review throughout the week. They come from all the sources listed above.

PostHog also has Reviews and Review Queues, which help you build these lists of traces and review them one by one.

Going through this process will help you and your team find new bad interactions and bugs to fix, which in turn leads to the creation of new evaluators to prevent regressions and detect other user inputs that still fail after your fix is live.

> ## Conclusion

Reviewer: I like descriptive conclusion titles better.
```suggestion
## Testing agents isn't one size fits all
```
